x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/EsqlBinaryComparison.java

```diff
@@ -118,9 +120,11 @@ protected EsqlBinaryComparison(
         Expression left,
         Expression right,
         BinaryComparisonOperation operation,
-        Map<DataType, EsqlArithmeticOperation.BinaryEvaluator> evaluatorMap
+        Map<DataType, EsqlArithmeticOperation.BinaryEvaluator> evaluatorMap,
+        EsqlArithmeticOperation.BinaryEvaluator nanosToMillisEvaluator,
+        EsqlArithmeticOperation.BinaryEvaluator millisToNanosEvaluator
     ) {
-        this(source, left, right, operation, null, evaluatorMap);
+        this(source, left, right, operation, null, evaluatorMap, nanosToMillisEvaluator, millisToNanosEvaluator);
```

Comment:
After consideration, passing in two special case evaluator factories seems like the least bad way to implement this. We need two functions because we need the ordering information for the operator, but also need to rely on parameter order to know which argument is milliseconds and which is nanoseconds. 

I thought about changing the evaluator map to be keyed by the ordered pair of input types, but that would add a lot of needless cruft to the common cases.  We could try something like using the single type keyed map if `commonType` returns a non-null value, and using a two keyed map if it doesn't, but that seemed like more complexity than necessary.

I'm open to other suggestions on how to manage dispatching these operations.